### PR TITLE
Update documentation around wavefunction

### PR DESCRIPTION
**Pull request by ye-luo**
## Proposed changes
I revisited the current documentation around wavefunction for derivatives for optimization.
I have two commits.
The first one updates the actual documentation content.
The second one recorder contests. backflow and estimator docs are moved after wavefunction. It causes massive diff unfortunately.
So please review on the first commit.

In addition, I caught a bug that the mass is not applied properly when taking kinetic energy derivatives and particles don't have equal mass.

## What type(s) of changes does this code introduce?
- Documentation changes

### Does this introduce a breaking change?
- No

## What systems has this change been tested on?
ryzen-box

## Checklist
- Yes. This PR is up to date with current the current state of 'develop'

### CODE REVIEW

File: docs/developing.rst [hunk 1]
```diff
@@ -2149,12 +2149,21 @@ An example of the second approach is
...
-``WaveFunctionComponent``. The code contains an example of a
+A full ``TrialWaveFunction`` is formulated as a product
+of all the components. Each component derives from ``WaveFunctionComponent``.
+
+.. math::
+     \psi = \prod_c {\tilde \psi_c}
+
+QMCPACK doesn't directly use the product form but mostly use the log of the wavefunction which
+fits the formulation of QMC algorithms very well. It also offers a numerical advantage on computers.
+The log value grows linearly instead of exponentioally with respect to the electron counts
```
Comment (markdewing) [inline docs/developing.rst]: The advantage of using the log vs. the value should be that the log doesn't exceed the range of double precision, where the value potentially could (??)  I'm not sure scaling wrt electron count matters, except that larger systems are more likely to suffer overflow/underflow on the value.

Comment (ye-luo) [reply docs/developing.rst]: by scaling I meant overflow/underflow on the value. Let me update.

Comment (ye-luo) [reply docs/developing.rst]: Fixed.

**Comment by ye-luo** [comment]
Test this please

**Comment by prckent** [comment]
From a casual reading this LGTM. Thanks @markdewing for the spelling corrections and typo fixes. Due to the understandable size of the diff I propose that we leave this open until late today to allow others to comment, then merge.

**Review by prckent — APPROVED** [review]

**Comment by prckent** [comment]
Start testing in-house

**Pull request merged**